tests/functional/test_metamodel/test_multi_metamodel_refs.py

```diff
@@ -0,0 +1,346 @@
+    mm_B = metamodel_from_str(grammarB, global_repository=global_repo,
+                              referenced_metamodels=[mm_A])
+
+    global_repo_provider = scoping_providers.PlainNameGlobalRepo()
+    mm_B.register_scope_providers({"*.*": global_repo_provider})
+    mm_A.register_scope_providers({"*.*": global_repo_provider})
+
+    mA = mm_A.model_from_str('''
+    A a1 A a2 A a3
+    ''')
+    global_repo_provider.add_model(mA)
+
+    _ = mm_B.model_from_str('''
+    B b1 -> a1 B b2 -> a2 B b3 -> a3
+    ''')
+
+    with raises(textx.exceptions.TextXSemanticError,
+                match=r'.*UNKNOWN.*'):
+        _ = mm_B.model_from_str('''
+        B b1 -> a1 B b2 -> a2 B b3 -> UNKNOWN
+        ''')
+
+
+def test_multi_metamodel_references2():
+    mm_A = metamodel_from_str(grammarA)
+    mm_B = metamodel_from_str(grammarB,
+                              referenced_metamodels=[mm_A])
+
+    global_repo_provider = scoping_providers.PlainNameGlobalRepo()
+    mm_B.register_scope_providers({"*.*": global_repo_provider})
+
+    mA = mm_A.model_from_str('''
+    A a1 A a2 A a3
+    ''')
+    global_repo_provider.add_model(mA)
+
+    _ = mm_B.model_from_str('''
+    B b1 -> a1 B b2 -> a2 B b3 -> a3
+    ''')
+
+    with raises(textx.exceptions.TextXSemanticError,
+                match=r'.*UNKNOWN.*'):
+        _ = mm_B.model_from_str('''
+        B b1 -> a1 B b2 -> a2 B b3 -> UNKNOWN
+        ''')
+
+
+def test_multi_metamodel_references_with_importURI():
+    # Use a global repo.
+    # This is useful, especially with circular includes or diamond shaped
+    # includes. Without such a repo, you might get double instantiations of
+    # model elements.
+    # However, if B includes A, but A not B, both meta models might have
+    # global repos on their own (global between model files of the same
+    # meta model --> global_repository=True). Circular dependencies
+    # will require shared grammars, like in test_metamodel_provider3.py,
+    # because it is not possible to share meta models for referencing, before
+    # the meta model is constructed (like in our example, mm_A cannot
+    # reference mm_B, if mm_B already references mm_A because one has to
+    # constructed first).
+    global_repo = scoping.GlobalModelRepository()
+
+    # Create two meta models with the global repo.
+    # The second meta model allows referencing the first one.
+    mm_A = metamodel_from_str(grammarA, global_repository=global_repo)
+    mm_B = metamodel_from_str(grammarBWithImport,
+                              global_repository=global_repo,
+                              referenced_metamodels=[mm_A])
+
+    # define a default scope provider supporting the importURI feature
+    mm_B.register_scope_providers({"*.*": scoping_providers.FQNImportURI()})
+
+    # map file endings to the meta models
+    scoping.MetaModelProvider.clear()
+    scoping.MetaModelProvider.add_metamodel("*.a", mm_A)
+    scoping.MetaModelProvider.add_metamodel("*.b", mm_B)
```

Comment:
Could we make this API a bit simpler for the user? If functions of `MetaModelProvider` are static we could add additional convenience function `add_metamodel` in the `scoping` module. So user could call `scoping.add_metamodel` in addition to a more verbose way as it is now. Also, we could add `scoping.clear_metamodels`. I think this could make it a bit easier to use and to remember.

Reply:
Hmm... Yes, I postpone this API change. 

The static methods are definitely a weak point (global variable) and further establish them by adding the proposed methods will not make it better.

I will maybe think about if I find a better (more local) solution without littering the global namespace...